src/main/java/codechicken/nei/BookmarkPanel.java

```diff
@@ -177,29 +177,40 @@ private void generateToDoSpaces() {
             }
 
             BookmarkStackMeta meta;
+            Boolean isFirstColumn = false;
+            Boolean isIngredient = false;
             int size = size();
             int idx = 0;
 
             while (idx < size) {
+                int lastIdx = idx;
                 ArrayList<Integer> pmask = new ArrayList<>();
 
                 for (int i = 0, perPage = rows * columns; i < perPage && idx < size; i++) {
+
                     if (isInvalidSlot(i)) {
                         pmask.add(null);
                     } else {
                         meta = getMetadata(idx);
+                        isFirstColumn = (i % columns) == 0;
+                        isIngredient = meta.recipeId != null && meta.ingredient;
 
-                        if (idx > 0 && ((i % columns) == 0) == (meta.recipeId != null && meta.ingredient
-                                && meta.recipeId.equals(getRecipeId(idx - 1)))) {
-                            pmask.add(null);
-                        } else {
+                        if (isFirstColumn && !isIngredient) {
                             pmask.add(idx++);
 
-                            if ((i % columns) == 0) {
-                                todoItemsCount++;
-                            }
+                            todoItemsCount++;
+                        } else if (isIngredient && (!isFirstColumn || i + 1 < perPage && isInvalidSlot(i + 1))) {
```

Comment:
Can you add a comment explaining the purpose of these checks so it's easier to read?

Reply:
done